finetuner/tuner/base.py

```diff
@@ -1,40 +1,107 @@
 import abc
 import warnings
-from typing import (
-    Generator,
-    Optional,
-    Union,
-    Tuple,
-    List,
-    Dict,
-)
-
-from ..helper import AnyDNN, AnyTensor, AnyDataLoader, AnyOptimizer, DocumentArrayLike
+from typing import Callable, Dict, Generic, List, Optional, Tuple, TypeVar, Union
+
+from .dataset import ClassDataset, SessionDataset
+from .dataset.samplers import RandomClassBatchSampler, SessionBatchSampler
 from .summary import Summary
+from ..helper import AnyDataLoader, AnyDNN, AnyOptimizer, AnyTensor, DocumentSequence
+
+
+LabelType = TypeVar('LabelType')
+
+
+class BaseMiner(abc.ABC, Generic[AnyTensor, LabelType]):
+    @abc.abstractmethod
+    def mine(self, labels: LabelType, distances: AnyTensor) -> Tuple[AnyTensor, ...]:
+        """Generate mined tuples from labels and item distances.
+
+        :param labels: labels of each item
+        :param distances: A tensor matrix of pairwise distance between each two item
+            embeddings
+
+        :return: A tuple of 1D tensors, denoting indices or properties of returned
+            tuples
+        """
+
+
+class BaseClassMiner(BaseMiner[AnyTensor, AnyTensor], Generic[AnyTensor]):
+    @abc.abstractmethod
+    def mine(self, labels: AnyTensor, distances: AnyTensor) -> Tuple[AnyTensor, ...]:
+        """Generate mined tuples from labels and item distances.
+
+        :param labels: A 1D tensor of item labels (classes)
+        :param distances: A tensor matrix of pairwise distance between each two item
+            embeddings
+
+        :return: A tuple of 1D tensors, denoting indices or properties of returned
+            tuples
+        """
+
+
+class BaseSessionMiner(
+    BaseMiner[AnyTensor, Tuple[AnyTensor, AnyTensor]], Generic[AnyTensor]
+):
+    @abc.abstractmethod
+    def mine(
+        self, labels: Tuple[AnyTensor, AnyTensor], distances: AnyTensor
+    ) -> Tuple[AnyTensor, ...]:
+        """Generate mined tuples from labels and item distances.
+
+        :param labels: A tuple of 1D tensors, denotind the items' session and match
+            type (0 for anchor, 1 for postive match and -1 for negative match),
+            respectively
+        :param distances: A tensor matrix of pairwise distance between each two item
+            embeddings
+
+        :return: A tuple of 1D tensors, denoting indices or properties of returned
+            tuples
+        """
 
 
 class BaseLoss:
```

Comment:
I think this designs does not impose any abstract method to BaseLoss, it seems that to be a valid loss u just need to provide a way to get default miner, but We should add a method `compute` or something, that in the `Pytorch` it should go to `forward`. Or maybe a `__call__` interface, but this BaseLoss is a poor interface

Reply:
Ok, let me think about it. using `__call__` won't work because then we mess with pytorch's implementation

Reply:
@JoanFM  I've added a `compute` function to the base loss class